### PR TITLE
ENH: Better error for rtree dependency in sjoin

### PR DESCRIPTION
**Pull request by rdmolony**
Following on from discussion in #1424 I've attempted to implement suggested changes.  This implementation may not be ideal so please feel free to ignore if not up to standard.

(see new error message at bottom)

---

# Example

```python
from shapely.geometry import Polygon
import geopandas as gpd

coords1 = ((0, 0), (0, 5), (2.5, 2.5))
gdf1 = gpd.GeoDataFrame({'geometry': [Polygon(coords1)]})
coords2 = ((1, 0), (0, 6), (2.5, 2.5))
gdf2 = gpd.GeoDataFrame({'geometry': [Polygon(coords2)]})

gpd.sjoin(gdf1, gdf2)
```
# Old error message for rtree

```python-traceback

/home/wsl-rowanm/miniconda3/envs/elec/lib/python3.8/site-packages/geopandas/base.py:104: UserWarning: Cannot generate spatial index: Missing package `rtree`.
  warn("Cannot generate spatial index: Missing package `rtree`.")
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/mnt/c/users/rowanm/documents/codema_elec/src/codema_elec/preprocess/test.py in <module>
      7 gdf2 = gpd.GeoDataFrame({'geometry': [Polygon(coords2)]})
      8 
----> 9 gpd.sjoin(gdf1, gdf2)

~/miniconda3/envs/elec/lib/python3.8/site-packages/geopandas/tools/sjoin.py in sjoin(left_df, right_df, how, op, lsuffix, rsuffix)
    133     else:
    134         # tree_idx_df == 'left'
--> 135         idxmatch = right_df.geometry.apply(lambda x: x.bounds).apply(
    136             lambda x: list(tree_idx.intersection(x)) if not x == () else []
    137         )

~/miniconda3/envs/elec/lib/python3.8/site-packages/pandas/core/series.py in apply(self, func, convert_dtype, args, **kwds)
   3846             else:
   3847                 values = self.astype(object).values
-> 3848                 mapped = lib.map_infer(values, f, convert=convert_dtype)
   3849 
   3850         if len(mapped) and isinstance(mapped[0], Series):

pandas/_libs/lib.pyx in pandas._libs.lib.map_infer()

~/miniconda3/envs/elec/lib/python3.8/site-packages/geopandas/tools/sjoin.py in <lambda>(x)
    134         # tree_idx_df == 'left'
    135         idxmatch = right_df.geometry.apply(lambda x: x.bounds).apply(
--> 136             lambda x: list(tree_idx.intersection(x)) if not x == () else []
    137         )
    138         idxmatch = idxmatch[idxmatch.apply(len) > 0]

AttributeError: 'NoneType' object has no attribute 'intersection'
```

# New error message for rtree
```python-traceback

/Users/rowanm/Documents/cloned-code/pandapower/geopandas_rdmolony/geopandas/sindex.py:38: UserWarning: Spatial indexes require either `rtree` or `pygeos`.
  warn("Spatial indexes require either `rtree` or `pygeos`.")
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-1-de3cd84fb5a1> in <module>
      7 gdf2 = gpd.GeoDataFrame({'geometry': [Polygon(coords2)]})
      8 
----> 9 gpd.sjoin(gdf1, gdf2)

~/Documents/cloned-code/pandapower/geopandas_rdmolony/geopandas/tools/sjoin.py in sjoin(left_df, right_df, how, op, lsuffix, rsuffix)
     80     package='rtree'
     81     if find_spec(package) is None:
---> 82         raise ImportError(
     83             f"{package} must be installed to use sjoin\n\n"
     84             "See installation instructions at https://geopandas.org/install.html"

ImportError: rtree must be installed to use sjoin

See installation instructions at https://geopandas.org/install.html
```

### CODE REVIEW

File: geopandas/tools/sjoin.py [hunk 1]
```diff
@@ -75,6 +76,15 @@ def sjoin(
             " joined".format(index_left, index_right)
         )
 
+    # Check if rtree is installed...
+    package='rtree'
+    if find_spec(package) is None:
+        raise ImportError(
+            f"{package} must be installed to use sjoin\n\n"
+            "See installation instructions at https://geopandas.org/install.html"
```
Comment (jorisvandenbossche) [inline geopandas/tools/sjoin.py:84]: You can reuse the `HAS_RTREE` variable from https://github.com/geopandas/geopandas/blob/master/geopandas/_compat.py here

File: setup.py [hunk 2]
```diff
@@ -29,7 +29,7 @@
 if os.environ.get("READTHEDOCS", False) == "True":
     INSTALL_REQUIRES = []
 else:
-    INSTALL_REQUIRES = ["pandas >= 0.23.0", "shapely", "fiona", "pyproj >= 2.2.0"]
+    INSTALL_REQUIRES = ["pandas >= 0.23.0", "shapely", "fiona", "pyproj >= 2.2.0",]
```
Comment (jorisvandenbossche) [inline setup.py:32]:
```suggestion
    INSTALL_REQUIRES = ["pandas >= 0.23.0", "shapely", "fiona", "pyproj >= 2.2.0"]
```

**Comment by jorisvandenbossche** [comment]
I am also adding a CI build with no optional dependencies in https://github.com/geopandas/geopandas/pull/1426, so we can actually test this error message

File: geopandas/tools/sjoin.py (hunk 1 above)
Comment (rdmolony) [reply geopandas/tools/sjoin.py:82]: Have swapped em out

**Comment by rdmolony** [comment]
Am a Cl build noob, only checks so far have been a brute force `conda install rtree` and `conda uninstall rtree`, curious to see how Cl build test is done

File: geopandas/tools/sjoin.py [hunk 3]
```diff
@@ -75,6 +75,13 @@ def sjoin(
             " joined".format(index_left, index_right)
         )
 
+    # Check if rtree is installed...
+    if not HAS_RTREE:
+        raise ImportError(
+            f"Rtree must be installed to use sjoin\n\n"
```
Comment (martinfleis) [inline geopandas/tools/sjoin.py:81]: We still support python 3.5, so no f-strings.

Comment (rdmolony) [reply geopandas/tools/sjoin.py]: Done (wasn't needed anyway), cheers @martinfleis

**Comment by martinfleis** [comment]
> Am a Cl build noob, only checks so far have been a brute force conda install rtree and conda uninstall rtree, curious to see how Cl build test is done

There will be a CI build (#1426) running without rtree, so you just want to check that it raises the error in that case. Something like this:

```py
if not HAS_RTREE:
    with pytest.raises(ImportError):
        geopandas.sjoin(...)
```

File: geopandas/tools/tests/test_sjoin.py [hunk 4]
```diff
@@ -7,10 +7,14 @@
...
 import geopandas
 from geopandas import GeoDataFrame, GeoSeries, read_file, sindex, sjoin
+from geopandas._compat import HAS_RTREE
 
 from pandas.testing import assert_frame_equal
 import pytest
 
+@pytest.fixture()
+def has_rtree():
+    return HAS_RTREE
```
Comment (martinfleis) [inline geopandas/tools/tests/test_sjoin.py:17]: I don't think we need a fixture here.

File: geopandas/tools/tests/test_sjoin.py [hunk 5]
```diff
@@ -84,6 +88,12 @@ def dfs(request):
 
     return [request.param, df1, df2, expected]
 
+@pytest.mark.parametrize("dfs", ["default-index", "string-index"], indirect=True)
+def test_raises_error_if_rtree_not_install(has_rtree, dfs): 
+    _, df1, df2, _ = dfs
+    has_rtree=False
+    with pytest.raises(ImportError):
+        geopandas.sjoin(df1, df2)
```
Comment (martinfleis) [inline geopandas/tools/tests/test_sjoin.py:96]:
```suggestion
    if not HAS_RTREE:
        with pytest.raises(ImportError):
            geopandas.sjoin(df1, df2)
```

**Comment by rdmolony** [comment]
Thanks @martinfleis for all the hand-holding (this is my first PR to a big project), I've merged this PR with master but don't know how to run this test only in a specific environment...  

Had a look at .travis.yml and travis resources: [link](https://docs.travis-ci.com/user/for-beginners) & [link](https://docs.travis-ci.com/user/customizing-the-build),  is a specific line in 'script' in .travis.yml required for this?

**Comment by adriangb** [comment]
If you're testing locally, we just use pytest. It's going to be a _lot_ easier to just manually manage the dependencies inside a virtualenv and install/uninstall to emulate the CI run than to get Travis to run locally.

Also cross referencing #1421, since after that is merged we would need to check for _any_ spatial index, not rtree specifically.

Personally, I would like to make changes to the sindex API so that if you try to create a spatial index (i.e. call `geopandas.sindex.get_sindex_class`) without any spatial index backend, you get an error instead of a warning.

**Comment by martinfleis** [comment]
> Also cross referencing #1421, since after that is merged we would need to check for any spatial index, not rtree specifically.
> Personally, I would like to make changes to the sindex API so that if you try to create a spatial index (i.e. call geopandas.sindex.get_sindex_class) without any spatial index backend, you get an error instead of a warning.

@adriangb Let's leave that for a follow-up PR. 

@rdmolony AppVeyor is failing due to unrelated issue, so feel free to ignore it. Travis is failing due to style issues, which means that files do not follow `black` formatting. You should ideally install [black](https://github.com/psf/black) and run `black geopandas` to fix that. And don't forget to remove that `pytest.fixture` :).

File: geopandas/tools/tests/test_sjoin.py [hunk 6]
```diff
@@ -7,11 +7,17 @@
 
 import geopandas
 from geopandas import GeoDataFrame, GeoSeries, read_file, sindex, sjoin
+from geopandas._compat import HAS_RTREE
 
 from pandas.testing import assert_frame_equal
 import pytest
 
 
+@pytest.fixture()
```
Comment (martinfleis) [inline geopandas/tools/tests/test_sjoin.py:16]: It is still here :). Can you remove lines 16-20 from this PR? These are not needed. Apart from that it is ready to be merged, CI is green and this new test is passing as it should.

Comment (rdmolony) [reply geopandas/tools/tests/test_sjoin.py]: Apologies! (forgot to save this first time around...)

**Comment by martinfleis** [comment]
Thank you @rdmolony! 

(I just pushed a fix removing the fixture from the test function itself.)

**Comment by rdmolony** [comment]
We got there in the end!  (so many mistakes...) Thanks @martinfleis

**Comment by rdmolony** [comment]
Sorry @martinfleis one last q (taken up far too much of your time), I noticed that last commit has  'Co-authored-by: = <user.email>'.  No need to change anything but in future how would you recommend co-committing?  

I found [this Github resource](https://help.github.com/en/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors) on co-authored commits, it recommends:

- GUI: using 'Github Desktop' which has +person option for co-commits
- Command line: manually writing 'Co-authored by:...' which seems a little tedious

(I'm happy to add this to the [geopandas' contribution guide](https://geopandas.readthedocs.io/en/latest/contributing.html) if you think this would be a good idea.  Also read through [pandas contribution guide](https://pandas.pydata.org/pandas-docs/stable/development/contributing.html#updating-your-pull-request) - should have done before - and found commit message convention that geopandas uses which could copy nto [geopandas' contribution guide](https://geopandas.readthedocs.io/en/latest/contributing.html) if desired)

**Comment by martinfleis** [comment]
> 'Co-authored-by: = <user.email>'.

Hmm, that is probably caused by my last commit. I have to investigate my settings, I've switched to VS Code recently so there might still be stuff to setup.

Co-authoring is usually quite straightforward as it get automatically linked if you directly accept suggestions or if we commit changes directly. This is not an issue.

If you'd like to update contributing guide based on pandas one, you are welcome to do so. Especially the section on `Updating your pull request` is helpful as we often ask contributors to do that.

**Pull request merged**